src/crypto.ts

```diff
@@ -0,0 +1,361 @@
+ * the provider is loaded, the wrapper calls `getHash` on the provider to obtain a real HashFunction instance and
+ * resolves any outstanding calls. Future calls on the wrapper are passed directly to the created HashFunction instance.
+ *
+ * When the HashFunction wrapper's `create` method is called, it returns a wrapped Hasher instance. Like the
+ * HashFunction wrapper, the Hasher wrapper will defer calls until a provider is loaded and a real Hasher implementation
+ * has been created, at which point all future calls to methods on the wrapper are passed directly to the real Hasher.
+ *
+ * The current provider may be requested with the `getProvider` function. Since the initial provider is loaded
+ * asynchronously, this function returns a Promise<CryptoProvider>. The provider may be set using the `setProvider`
+ * function, which accepts a CryptoProvider.
+ */
+
+import Promise from 'dojo-core/Promise';
+import global from 'dojo-core/global';
+import has, { add as hasAdd } from 'dojo-core/has';
+import { Sink } from 'dojo-core/streams/WritableStream';
+
+hasAdd('webcrypto', typeof global.SubtleCrypto !== 'undefined');
+
+declare var define: { amd: any };
+declare var require: Function;
+
+export type Binary = Uint8Array | Buffer | number[];
+// TODO: this should be an encoding Codec (or equivalent) when that exists
+export type Codec = string;
+export type Data = string | Binary;
+
+/**
+ * An interface describing a cryptographic provider.
+ */
+export interface CryptoProvider {
+	getHash(algorithm: string): HashFunction;
+	getSign(algorithm: string): SignFunction;
+}
+
+/**
+ * The current provider. Providers supply concrete implementations of the API described here. Users should not typically
+ * need to access providers directly.
+ */
+let provider: Promise<CryptoProvider>;
+
+/**
+ * Gets the HashFunction for a particular algorithm. The algorithm is specified as a string for simplicity and
+ * extensibility.
+ */
+export function getHash(algorithm: string): HashFunction {
+	let realHash: HashFunction;
+	const hashPromise = new Promise<HashFunction>(function (resolve, reject) {
+		getProvider().then(
+			function (provider) {
+				try {
+					realHash = provider.getHash(algorithm);
+					resolve(realHash);
+				}
+				catch (error) {
+					reject(error);
+				}
+			},
+			function (error) {
+				reject(error);
+			}
+		);
+	});
+
+	// Return a wrapper that will defer calls to the hash until a provider has been loaded.
+	const hashFunction = <HashFunction> function (data: Data, codec?: Codec): Promise<Binary> {
+		if (realHash) {
+			return realHash(<any> data, codec);
+		}
+		return hashPromise.then<Binary>(function (hash) {
+			return hash(<any> data, codec);
+		});
+	}
+
+	// Return a wrapper class that will defer calls until a provider has been loaded and an actual Hasher instance has
+	// been created.
+	hashFunction.create = function<T extends Data> (codec?: Codec): Hasher<T> {
```

Comment:
Is it feasible to replace this once we know the actual hasher is loaded?  (I'm wondering whether `setProvider` prohibits this optimization, but I would think you're going to have to re-call `getHash` and `getSign` after calling `setProvider` anyway.)